### ISSUE TITLE
Sound Blaster Pro 1 and 2 are reversed in DOSBox

### ISSUE THREAD

**Issue by James-F2** [dosbox-staging/dosbox-staging#2942]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

dosbox-staging version 0.81.0-alpha-1676-gfed6c628d

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Sound Blaster Pro 1 and 2 PCM are reversed in DOSBox.

dosbox-staging version 0.81.0-alpha-1676-gfed6c628d

Several test in popular games.
https://www.vogons.org/viewtopic.php?p=536104#p536104

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by James-F2** [comment]
To add to the reversed output filter register here:
https://github.com/dosbox-staging/dosbox-staging/issues/2869

This makes the SBPro emulation quite buggy and not recommended (by me) to use.... yet.

**Comment by johnnovak** [comment]
Thanks for the report and the super informative linked thread @James-F2 👍🏻 I'll get to the bottom of it at some point. In the meantime, not sure if you're aware but can simply reverse the left and right channels with the `MIXER SB REVERSE` command.

**Comment by James-F2** [comment]
> In the meantime, not sure if you're aware but can simply reverse the left and right channels with the `MIXER SB REVERSE` command.

That will also reverse the CD audio and Mixer L/R channel which are correct.
The reversal is only for the interleaved stereo PCM stream, it happens in most games but not all.
Fixing this will require an in-depth research of the SBPro card, unfortunately this will not be a quick fix.

**Comment by johnnovak** [comment]
> > In the meantime, not sure if you're aware but can simply reverse the left and right channels with the `MIXER SB REVERSE` command.
> 
> That will also reverse the CD audio and Mixer L/R channel which are correct. The reversal is only for the interleaved stereo PCM stream, it happens in most games but not all. Fixing this will require an in-depth research of the SBPro card, unfortunately this will not be a quick fix.

Cool I'll read the whole thing, but `CDAUDIO` has its own channel. The `SB` channel is only for PCM audio and nothing else.

**Comment by James-F2** [comment]
> Cool I'll read the whole thing, but `CDAUDIO` has its own channel. The `SB` channel is only for PCM audio and nothing else.

Got it, I'll try that.

Another temporary solution to have a good SBPro2 experience with correct stereo:
sbtype              = sb16
sb_filter           = sbpro2
opl_filter          = sbpro2

**Comment by johnnovak** [comment]
Found a [super informative post](https://www.vogons.org/viewtopic.php?p=943606#p943606) on the last page of the linked thread which seems to be the definitive answer on the issue. Sample C code for a very comprehensive testing tool is also included.

**Comment by James-F2** [comment]
Yep.
I think he talks about what's going on internally in the DSP, but from the perspective of the actual games and what's coming out of the speakers, the real SBPro is not reversed because the game developers compensated for that, whether DOSBox does not.
In any case, the linked thread has plenty of info and tools to correct the emulation but I am not a programmer and have no idea how hard it would be.

**Comment by johnnovak** [comment]
@James-F2 yeah, the observable behaviour is crystal clear and you've documented that extremely well, thanks a lot for that!

Read the whole thing again, plus found another comment that rules out some theories: https://www.vogons.org/viewtopic.php?p=536908#p536908

This is actually in line with the earlier comment I referenced, and the solution appears to be:
- The PCM output of the SB Pro _is_ reversed from an implementation perspective (this is important)
- The very first 1 byte DMA transfer _after_ a DSP reset should be ignored
- Following odd sized DMA transfers shoul "carry over"
- CD Audio and the SB Mixer should _not_ be reversed
- This should give the observable "non-reversed" behaviour documented by @James-F2 

Also, as the commenter noted, PCem v17 has a perfect implementation of the stereo handling.

**Comment by kcgen** [comment]
Extremely useful summary, @johnnovak!

**Comment by johnnovak** [comment]
> Yep, great summary.
> 
> Wolfenstein 3D disables the filter even if it's forced in dosbox config file (sb_filter_always_on = true), even with SB16. I think it's related to "first 1 byte DMA transfer after a DSP reset should be ignored", but I'm not sure.
> 
> I do see "SB16: DSP was reset" when I start Wolfenstein 3D.

I think this in line with your previous observation about the filter behaviour: the DSP reset should probably not alter the filter state, and it seems we have the programmatic filter setting flipped on top of that. To be investigated as well.

**Comment by Grounded0** [comment]
I discovered that some or even many games with Miles Sound System will have reversed stereo by default when using SB16. Since people are looking "stereo" in our issue tracker first let me put a simple workaround here.

If the game doesn't have reverse stereo option that is stored across game sessions you can do a batch file like this:

```
MIXER SB REVERSE /NOSHOW
GAME.EXE
MIXER SB STEREO /NOSHOW
```

This will reverse stereo when the game is run and switch back to stereo once the game has stopped running.

**Comment by Grounded0** [comment]
@johnnovak 

You're stipulating we got both stereo and filter behaviour now right for Sound Blaster Pro here:

- #3815

Can this ticket be closed now?

**Comment by johnnovak** [comment]
> You're stipulating that we now got both stereo and filter behaviour correct for Sound Blaster Pro series here:

To clarify, when I said stereo behaviour, I meant the bit that toggles between stereo and mono operation. That is handled correctly.

This ticket is about the reversal of the left and right channels when the stereo bit is enabled — different issue, and still outstanding.

**Comment by James-F2** [comment]
Yep, SBPro PCM is still reversed in 82.0 RC1.

**Comment by Grounded0** [comment]
We'll close this once its resolved.

**Comment by weirddan455** [comment]
I read through the Vogons post and did some testing of my own. I believe @johnnovak hit the nail on the head with the proposed solution. PR in #4412 Hopefully this can finally close out this 9 year old bug 😄 

>     * The PCM output of the SB Pro _is_ reversed from an implementation perspective (this is important)

Yes, the SoundBlaster Pro hardware appears to always reverse the channels compared to the SoundBlaster 16. Games compensate for this by reversing the channels again in software if they detect the SBPro (confirmed this is exactly what Doom does).

>     * The very first 1 byte DMA transfer _after_ a DSP reset should be ignored

This is odd behavior but it also appears to be true. This lets DOSBox match real hardware for Quake and SBTEST.EXE

>     * Following odd sized DMA transfers should "carry over"

Yes, it's unusual for software to use oddly sized DMA buffers when playing Stereo but at least the demo Panic by Future Show depends on the dangling samples to carry over to the next transfer. Otherwise you get audio channels reversing multiple times at runtime and some small stutters.

We were already handling this correctly but this behavior needs to stay.

>     * CD Audio and the SB Mixer should _not_ be reversed

I assume you mean OPL should not be reversed. I'm not doing anything at the mixer level anyway in my PR.

**Comment by johnnovak** [comment]
> I read through the Vogons post and did some testing of my own. I believe [@johnnovak](https://github.com/johnnovak) hit the nail on the head with the proposed solution. PR in [#4412](https://github.com/dosbox-staging/dosbox-staging/pull/4412) Hopefully this can finally close out this 9 year old bug 😄

Better late than never, and it's a confusing issue, indeed! 😄 It's great you've been able to confirm my summary is correct. The big thanks goes to the guy in the linked Vogons post who reversed the actual hardware behaviour (90% of the other posts were just random theories that only obfuscated the issue further).

> I assume you mean OPL should not be reversed. I'm not doing anything at the mixer level anyway in my PR.

Yes, what I was trying to say only the SB DAC producing the stereo stream should be affected by the fix that is then fed into the SB mixer.
